[imgui] #ifndef guards for possibly already set definitions

VelocityRa opened the pull request:
We already set them in `CMakeLists.txt`, resulting in a macro re-definition compiler warning. Instead of removing these `vcpkg_replace_string` calls entirely which would _probably_ work with no side-effects, I chose to add these guards in case a user wants to manually modify `imconfig.h` or read which options are enabled.

<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->

<!-- If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/ -->

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [x] The "supports" clause reflects platforms that may be fixed by this new version
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

<!-- If this PR adds a new port, please uncomment and fill out this checklist:

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [ ] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [ ] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [ ] The versioning scheme in `vcpkg.json` matches what upstream says.
- [ ] The license declaration in `vcpkg.json` matches what upstream says.
- [ ] The installed as the "copyright" file matches what upstream says.
- [ ] The source code of the component installed comes from an authoritative source.
- [ ] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is in the new port's versions file.
- [ ] Only one version is added to each modified port's versions file.

END OF NEW PORT CHECKLIST (delete this line) -->

VelocityRa commented:
I can't repro the issue locally, it all works for me, including if I use the wchar32 feature.

How can I look at logs like `D:\b\milerius-sfml-imgui\install-x86-windows-dbg-out.log` from CI so I can debug this.

cc @LilyWangLL

LilyWangLL commented:
You could download the error logs from https://dev.azure.com/vcpkg/public/_build/results?buildId=97899&view=artifacts&pathAsName=false&type=publishedArtifacts.

VelocityRa commented:
@LilyWangLL Thanks. The issue is this:
```
imconfig.h:85:2: error: unterminated conditional directive
#ifndef IMGUI_ENABLE_FREETYPE
```
but I can't figure out why.

Isn't this correct?
```cmake
    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/include/imconfig.h" "//#define IMGUI_ENABLE_FREETYPE"
        "#ifndef IMGUI_ENABLE_FREETYPE\n#define IMGUI_ENABLE_FREETYPE\n#endif")
```
I see some similar code in ie. https://github.com/microsoft/vcpkg/blob/f1c6efee2245009540dde947e0e3d008f3aa7dbb/ports/dlib/portfile.cmake#L72 too.

LilyWangLL commented:
> @LilyWangLL Thanks. The issue is this:
> 
> ```
> imconfig.h:85:2: error: unterminated conditional directive
> #ifndef IMGUI_ENABLE_FREETYPE
> ```
> 
> but I can't figure out why. Seems to be missing the `#endif` for some reason?
> 
> Isn't this correct?
> 
> ```cmake
>     vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/include/imconfig.h" "//#define IMGUI_ENABLE_FREETYPE"
>         "#ifndef IMGUI_ENABLE_FREETYPE\n#define IMGUI_ENABLE_FREETYPE\n#endif")
> ```
> 

The results after this change:
![image](https://github.com/microsoft/vcpkg/assets/94091114/705b01f6-b114-471d-b28f-124dab14d147)

Please see line 87, this is the reason of this error.

VelocityRa committed:
[imgui] #ifndef guards for possibly already set definitions

We already set them in CMakeLists.txt, resulting in macro re-definition
compiler warnings.  Instead of removing these vcpkg_replace_string
calls entirely, we choose to add these guards, in case a user wants to
manually modify imconfig.h or read which options are enabled.

VelocityRa commented:
@LilyWangLL Thank you, fixed.

BillyONeal commented:
> We already set them in `CMakeLists.txt`, resulting in a macro re-definition compiler warning.

Can you show how to reproduce this problem?

LilyWangLL commented:
Pinging @VelocityRa for response. Can you address the review suggestions? Thanks!

LilyWangLL commented:
losing this PR since it seems that no progress is being made. Please ping us to reopen if work is still being done.
